pareto/utilities/bounding_functions.py

```diff
@@ -0,0 +1,161 @@
+    elif model.config.water_quality is WaterQuality.false:
+        pass
+    else:
+        pass
+    return model
+
+
+###############################################################################
+### Individual variable bound generation functions
+def Bound_v_F_Piped(model):
+    # initialization function for parameter
+    def p_F_Piped_UB_init(model, l, l_tilde):
+        # set default value to pieline capacity plus maximum expansion capacity
+        return model.p_sigma_Pipeline[l, l_tilde] + max(
+            [value(model.p_delta_Pipeline[d]) for d in model.s_D]
+        )
+
+    # Set up bounds in a parameter
+    model.p_F_Piped_UB = Param(
+        model.s_L,
+        model.s_L,
+        default=None,
+        mutable=True,
+        within=Any,
+        initialize=p_F_Piped_UB_init,
+        units=model.model_units["volume_time"],
+        doc="Maximum pipeline capacity between nodes [volume_time]",
+    )
+    # Assign upper bounds to variables; note that using a bounds rule would require redefining the variable
+    for key in model.s_LLA:
+        for t in model.s_T:
+            model.v_F_Piped[key, t].setub(model.p_F_Piped_UB[key])
+    return model
+
+
+def Bound_v_F_Trucked(model):
+    for key in model.s_LLT:
+        for t in model.s_T:
+            model.v_F_Trucked[key, t].setub(
+                model.p_delta_Truck * model.p_max_number_of_trucks
+            )
+    return model
+
+
+def Bound_v_Q_PP(model):
+    """
+    Variable bounding function for water quality variable v_Q in the post
+    processing strategic water quality extension
+
+    Notes on contents:
+        - uses the maximum value among several parameters at each node as bound
+        - not all parameters will be defined on every index
+        - "None" is a valid entry for a pyomo variable bound, meaning no bound
+        - function "TryOrNone()" is used to return None if a parameter is not
+          defined on a given index
+        - If no parameter is defined at an index, then use None as that bound
+    """
+    quality_highest_bound = max(
+        max([x.value for x in model.quality.p_xi_StorageSite.values()]),
+        max([x.value for x in model.quality.p_xi_PadStorage.values()]),
+        max([x.value for x in model.quality.p_nu_pad.values()]),
+        max([x.value for x in model.quality.p_nu_freshwater.values()]),
+    )
+
+    # Create a parameter with maximum quality values
+    model.quality.p_Q_UB = Param(
+        model.quality.s_WQL,
+        model.s_QC,
+        within=Any,
+        default=None,
+        mutable=True,
+        # initialize=p_Q_UB_init,
+        units=model.model_units["concentration"],
+    )
+    # Assign upper bounds to variables
+    for l in model.quality.s_WQL:
+        for w in model.s_QC:
+            for t in model.s_T:
+                # Individual quality bounds (NOTE: HAVE SHOWN, THIS DOES NOT WORK)
```

Comment:
What does the "NOTE" here mean?

Reply:
I tested the commented line that follows on 115 where the bound is derived from the p_Q_UB parameter. The logic that is used to create that parameter is not correct, resulting in a bound that is too tight, and the model solution will change with that bound enforced. I kept it there for two reasons: one, so it is available for testing should the same idea be tried in future, and two, as a record that the approach did not function properly. The current bound is fairly loose, and there is room for improvement in future versions.

Reply:
@Tominapa that seems reasonable to me, but I think you might as well comment out all of the code for `model.quality.p_Q_UB` (lines 101-109) for now, since it is currently unused.

Reply:
Commented